oeps/oep-0005.rst

```diff
@@ -0,0 +1,182 @@
+   has to be updated by every user after downloading (to update to the latest
+   code versions), or the image has to be rebuilt every time any of the installed
+   services change (which results in lots of churn).
+6. Because the `Vagrant`_ image contains many separate services, the definition
+   of that image has to be separate from any of the service repositories, which
+   makes it hard to update (and to push updates out to users).
+
+
+Specification
+=============
+
+Interface
+~~~~~~~~~
+
+The primary developer interaction with ``devstack`` will be through ``make``
+shortcuts. The following targets will be standard::
+
+    make devstack.watch  # Start the devstack services (if they aren't already)
+                         # and display their logs on stdout
+
+    make devstack.ssh    # Start the devstack services (if they aren't already)
+                         # and ssh into the primary service (as the user that
+                         # owns the service, in the directory that has the
+                         # service source code)
+
+    make devstack.clean  # Stop the devstack services and remove any intermediate
+                         # data that they stored in the .devstack/ directory
+
+    make devstack.*      # Start the devstack services (if they aren't already)
+                         # and run `make *` from the code directory of the service
+                         # (as the service owner)
+
+Implementation
+~~~~~~~~~~~~~~
+
+Requirements
+------------
+
+1. Services that aren't part of the Open edX ecosystem must be run using
+   their own published containers (where available).
+2. Open edX containers must run all of their processes via `supervisor`_ to
+   provide a single entrypoint even when the container has multiple processes
+   (such as `nginx`_ and a `Django`_ service running under `gunicorn`_).
+3. Open edX services running in containers must log to ``stdout`` and ``stderr``
+   so that those logs can be captured by `supervisor`_.
+4. Dependent services (such as databases) should be configured to mount their
+   storage directories in ``.devstack/`` inside the source code directory (this
+   should be added to `.gitignore` and `.dockerignore`_. This will allow data from
+   ``devstack`` to be persist across container executions.
+5. The Open edX service must mount the source code directory from the host
+   machine into the ``devstack`` container so that ``devstack`` is always running
+   the version of code being edited.
+6. ``devstack`` services should be configured to auto-reload when source code
+   is modified.
+7. Open edX Service containers should be built during CI, and published so that any
+   developer can download the latest container. They should also be tagged for
+   each versioned release.
+
+Technology Selection
+--------------------
+
+The prototype implementation was done using `Docker`_ for the containers, and
+`Docker Compose`_ to manage the networking between services.
```

Comment:
> We've endeavored to find an abstraction that supports multiple virtualization targets in ansible using he vhost and container roles -- which, parenthetically, was useful for the recent open stack changes.
> The abstraction is by-no-means perfect and it doesn't solve for the fact that some, important, differences will be in play comparing vagrant to aws to docker.

Yup, I agree that having multiple virtualization targets is a good thing, as it helps making the infrastructure less tied to a specific service (and specifically AWS, which is proprietary). The part I'm objecting to, it's to extend this support to pseudo-virtualization, which comes with a larger set of quirks of its own, and breaks assumptions that are common to the other full virtualization approaches.

But I also recognize that I'm in the minority here - so I mainly intended to mention my reservations, not to block the initiative.

> I'm not convinced that the mechanism for starting supervisord is likely to be a big source of risk. 

Yes, starting supervisord shouldn't be an issue - it's actually a point of similarity with the full virtualization approaches, which already use supervisord to run the services. (Although I always wondered why we weren't using the standard init scripts for this?)

> I take the point that not every service is managed by supervisor, but I think we can manage that by moving more things to supervisor, adopting 12 factor principles when sensible and minimizing customizations to "vendor" services.

That's still a significant overhead, no? And it also means we can't easily reuse standard ansible galaxy roles for standard services.

> To @cpennington's point, we can and do run multiple processes on Docker containers, say nginx, rsyslog and gunicorn, though I think that we should attempt, over time, to limit that.

You do that via supervisord?